<div class="alert alert-warning">

hw7_sol (6).ipynb
    
</div>

In [ ]:
top_topics = document_topics[:5].argmax(axis=1)

for i, topic_num in enumerate(top_topics):
    print(f"Document {i}:")
    print(f"Assigned Topic: {topic_num}")
    print(f"Text: {df['text'].iloc[i]}")
    print()

Document 0:
Assigned Topic: 3
Text: You know, I was reading 18 U.S.C. 922 and something just did not make 
sence and I was wondering if someone could help me out.

Say U.S.C. 922 :

(1) Except as provided in paragraph (2), it shall be unlawful for
any person to transfer or possess a machinegun.

 Well I got to looking in my law dictionary and I found that a "person" 
might also be an artificial entity that is created by government 
and has no rights under the federal constitution. So, what I 
don't understand is how a statute like 922 can be enforced on 
an individual. So someone tell me how my government can tell
me what I can or cannot possess. Just passing a law 
does not make it LAW. Everyone knows that laws are constitional
until it goes to court. So, has it ever gone to court, not
just your run of the mill "Ok I had it I am guilty, put me in jail"

Has anyone ever claimed that they had a right to possess and was told
by the Supreme Court that they didn't have that right?




Docu

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (32).ipynb
    
</div>

In [ ]:
doc_topic_dist = lda.transform(transformed[:5])
doc_topic_dist

array([[0.00250191, 0.00250412, 0.6736669 , 0.00250174, 0.31131755,
        0.00250205, 0.00250384, 0.00250188],
       [0.00176209, 0.00176258, 0.0017624 , 0.00176226, 0.00176247,
        0.45947738, 0.52994825, 0.00176256],
       [0.00255298, 0.00255495, 0.76823895, 0.00255543, 0.00255548,
        0.00255355, 0.21643463, 0.00255402],
       [0.00625411, 0.00625596, 0.00626404, 0.00625444, 0.1224402 ,
        0.00625883, 0.8400108 , 0.00626162],
       [0.13922576, 0.00368272, 0.00368269, 0.00367976, 0.83869089,
        0.00368051, 0.00367886, 0.00367881]])

In [ ]:
most_likely_topics = np.argmax(doc_topic_dist, axis=1)
for topic_id in most_likely_topics:
    print(newsgroups_train.target_names[topic_id])

comp.windows.x
talk.politics.guns
comp.windows.x
talk.politics.guns
rec.sport.hockey


The model is relatively confident in identifying the label for the first, third, fourth, and fifth document and deciding between two similar labels for the second document. This suggests that the model is efficient for topic modeling and the probability distribution returned gives an informative measure.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (51).ipynb
    
</div>

In [ ]:
document_topics[0]
first5 = document_topics[:5]
i = 0
for document in first5:
    print('\ndocument ' + str(i))
    print('Mideast/Atheism: ' + str(document[0]))
    print('Christianity:    ' + str(document[1]))
    print('Gun Politics:    ' + str(document[2]))
    print('Sports:          ' + str(document[3]))
    print('Computer:        ' + str(document[4]))
    print('Mideast:         ' + str(document[5]))
    i += 1


document 0
Mideast/Atheism: 0.0031686932703374644
Christianity:    0.0031698744747841387
Gun Politics:    0.9550627906753731
Sports:          0.003162485058950938
Computer:        0.0031678042136514175
Mideast:         0.03226835230690306

document 1
Mideast/Atheism: 0.002193174040462732
Christianity:    0.002187786856752882
Gun Politics:    0.12549203118658822
Sports:          0.0021899164260827334
Computer:        0.8657503965157255
Mideast:         0.002186694974388028

document 2
Mideast/Atheism: 0.0026796269976136432
Christianity:    0.002679355217030096
Gun Politics:    0.6466538852782806
Sports:          0.002666605874347156
Computer:        0.3426573067132607
Mideast:         0.0026632199194677

document 3
Mideast/Atheism: 0.09300991760694904
Christianity:    0.005626249579629453
Gun Politics:    0.005604175385136674
Sports:          0.005572524972849371
Computer:        0.8846093533824493
Mideast:         0.005577779072986216

document 4
Mideast/Atheism: 0.7586639804434232
Ch

In [ ]:
df.iloc[2]['text']

"\nActuallay I don't, but on the other hand I don't support the idea of having\none newsgroup for every aspect of graphics programming as proposed by Brian,\nin his reply to my original posting.\nI would suggest a looser structure more like a comp.graphics.programmer,\ncomp.graphics.hw_and_sw\nThe reason for making as few groups as possible is for the same reason you\nsay we shouldn't spilt up, not to get to few postings every day.\nI takes to much time to browse through all postings just to find two or \nthree I'm interested in.\n\nI understand and agree when you say you want all aspects of graphics in one\nmeeting. I agree to some extension. I see news as a forum to exchange ideas,\nhelp others or to be helped. I think this is difficult to achive if there\nare so many different things in one meeting.\n\nGood evening netters|-)"

Document 0: Gun Politics 95%

Document 1: Computer 86.6% and Gun politics 12.5%. This combination does not make much sense. The text is mainly talking about computers.

Document 2: Gun Politics 64.6% and Computer 34.3%. This combination does not make much sense. The text is mainly talking about computer graphics.

Document 3: Computer 88.4%

Document 4: Mideast 75% and Christianity 22.5%. The text is mainly talking about Mideast politics.

Most of the documents ended matching the correct topics other that Document 2. This is probably because the top 10 words for topic 2 was not present in the text but words that were in topic 2 appeared more often and were more generic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (20).ipynb
    
</div>

In [ ]:
print("1.")
for i in range(10):
    # print(df.iloc[i]['text'])
    print("Document " + str(i))
    print(document_topics[i])
print("2.")
print("The document topic assignment for Topic 0, 5-9 was very clear, with one or two majorities over the other. For instnace, Doc 0 had 80.86% topic 2 (starting from index 0), while the next largest contribution was 13.89% for topic 7.")
print("The rest of the documents had a mix of topics. In reference to the percentages, they represent the contribution of each topic to the document. By inspecting the text from each model, I would say the classification was appropriate and actually pretty accurate.")

1.
Document 0
[0.00142858 0.00142904 0.80861338 0.00142879 0.00142878 0.00142894
 0.00142862 0.13898052 0.04240466 0.00142869]
Document 1
[0.00129883 0.00129897 0.00129896 0.00129886 0.40728571 0.09504461
 0.00129873 0.45956497 0.00129905 0.03031132]
Document 2
[0.00149261 0.00149279 0.00149299 0.10696258 0.00149284 0.57651203
 0.00149258 0.30607611 0.00149289 0.0014926 ]
Document 3
[0.00256489 0.00256475 0.00256471 0.00256436 0.50032042 0.00256543
 0.00256415 0.47916228 0.00256483 0.00256419]
Document 4
[0.00217394 0.15033274 0.00217418 0.00217434 0.00217409 0.00217431
 0.00217394 0.00217424 0.51956726 0.31488095]
Document 5
[0.00344857 0.00344921 0.00344913 0.00344874 0.00344918 0.00344915
 0.00344847 0.59464443 0.37776473 0.00344839]
Document 6
[1.79541067e-04 1.79577835e-04 1.79584129e-04 1.79569127e-04
 1.79579410e-04 1.79589121e-04 1.79548352e-04 3.44806878e-01
 6.53756576e-01 1.79557177e-04]
Document 7
[0.01111114 0.0111145  0.01111785 0.01111476 0.56119148 0.01111379
 0.3498964

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (43).ipynb
    
</div>

In [ ]:
pd.DataFrame(document_topics).head()

0         1         2         3         4         5         6  \
0  0.063564  0.323205  0.002452  0.002452  0.002454  0.600966  0.002453   
1  0.499397  0.001764  0.001763  0.001763  0.383199  0.108587  0.001764   
2  0.002404  0.612585  0.002405  0.002405  0.372979  0.002408  0.002408   
3  0.004310  0.004317  0.004311  0.004311  0.667809  0.306312  0.004317   
4  0.002907  0.002915  0.002908  0.002911  0.002913  0.002911  0.002911   

          7  
0  0.002454  
1  0.001763  
2  0.002405  
3  0.004313  
4  0.979624

In [ ]:
pd.set_option('display.max_colwidth', None)
pd.DataFrame(df["text"]).head()

text
0                                                                                                                    You know, I was reading 18 U.S.C. 922 and something just did not make \nsence and I was wondering if someone could help me out.\n\nSay U.S.C. 922 :\n\n(1) Except as provided in paragraph (2), it shall be unlawful for\nany person to transfer or possess a machinegun.\n\n Well I got to looking in my law dictionary and I found that a "person" \nmight also be an artificial entity that is created by government \nand has no rights under the federal constitution. So, what I \ndon't understand is how a statute like 922 can be enforced on \nan individual. So someone tell me how my government can tell\nme what I can or cannot possess. Just passing a law \ndoes not make it LAW. Everyone knows that laws are constitional\nuntil it goes to court. So, has it ever gone to court, not\njust your run of the mill "Ok I had it I am guilty, put me in jail"\n\nHas anyone ever claimed that they had a right to possess and was told\nby the Supreme Court that they didn't have that right?\n\n\n
1  \n\n\nIt's not a bad question: I don't have any refs that list this algorithm\neither. But thinking about it a bit, it shouldn't be too hard.\n\n1) Take three of the points and find the plane they define as well as\nthe circle that they lie on (you say you have this algorithm already)\n\n2) Find the center  of this circle. The line passing through this center\nperpendicular to the plane of the three points passes through the center of\nthe sphere.\n\n3) Repeat with the unused point and two of the original points. This\ngives you two different lines that both pass through the sphere's\norigin. Their interection is the center of the sphere.\n\n4) the radius is easy to compute, it's just the distance from the center to\nany of the original points.\n\nI'll leave the math to you, but this is a workable algorithm. :-)\n\n\nAn alternate method would be to take pairs of points: the plane formed\nby the perpendicular bisector of each line segment pair also contains the\ncenter of the sphere. Three pairs will form three planes, intersecting\nat a point. This might be easier to implement.
2                                                                                                                                                                                                                                                                 \nActuallay I don't, but on the other hand I don't support the idea of having\none newsgroup for every aspect of graphics programming as proposed by Brian,\nin his reply to my original posting.\nI would suggest a looser structure more like a comp.graphics.programmer,\ncomp.graphics.hw_and_sw\nThe reason for making as few groups as possible is for the same reason you\nsay we shouldn't spilt up, not to get to few postings every day.\nI takes to much time to browse through all postings just to find two or \nthree I'm interested in.\n\nI understand and agree when you say you want all aspects of graphics in one\nmeeting. I agree to some extension. I see news as a forum to exchange ideas,\nhelp others or to be helped. I think this is difficult to achive if there\nare so many different things in one meeting.\n\nGood evening netters|-)
3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              The following problem is really bugging me,\na

Text 0: The most likely topic is topic 5 which I labeled as "Gun Laws" which seems pretty fitting.  
Text 1: The most likely topic is topic 0 which I labeled as "Gun Politics" which does NOT make sense here.  
Text 2: The most likely topic is topic 1 which I labeled as "Religion" which doesn't quite make sense here.  
Text 3: The most likely topic is topic 4 which I labeled as "Computer program" which seems quite fitting here.  
Text 4: The most liekly topic is topic 7 which I labeled as "Armenia - Turkey conflict" which seems fitting as well.

Overall it's not amazing, but from the first five examples the document topic is able to get 3/5 correct which is something at least!

<div class="alert alert-warning">

hw7_sol (16).ipynb
    
</div>

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (41).ipynb
    
</div>

In [ ]:
document_topic_dist = lda.transform(X[:5])

for i, doc_topics in enumerate(document_topic_dist):
    print(f"Document {i + 1}:")
    for topic, prob in enumerate(doc_topics):
        print(f"Topic {topic}: Probability {prob:.4f}")
    print()

Document 1:
Topic 0: Probability 0.0051
Topic 1: Probability 0.3862
Topic 2: Probability 0.6037
Topic 3: Probability 0.0050

Document 2:
Topic 0: Probability 0.8761
Topic 1: Probability 0.1165
Topic 2: Probability 0.0037
Topic 3: Probability 0.0037

Document 3:
Topic 0: Probability 0.2267
Topic 1: Probability 0.7631
Topic 2: Probability 0.0051
Topic 3: Probability 0.0051

Document 4:
Topic 0: Probability 0.4778
Topic 1: Probability 0.4999
Topic 2: Probability 0.0114
Topic 3: Probability 0.0110

Document 5:
Topic 0: Probability 0.0060
Topic 1: Probability 0.0062
Topic 2: Probability 0.9819
Topic 3: Probability 0.0059



This means that Document 1 is mostly Topic 2 (Politics) since it is 60% which is correct by the dataset, Document 2 is 87% Topic 0 (Computer), Document 3 is 76% Topic 1 (Religion) but it was supposed to be Computer, Document 4 is between Topic 0 (Computer) and Topic 1 (Religion) but it is supposed to be more Computer, Document 5 is 98% Topic 2 (Politics) as in the dataset.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (22).ipynb
    
</div>

In [ ]:
print(document_topics[0], document_topics[1], document_topics[2], 
      document_topics[3], document_topics[4])

[0.91792255 0.0048727  0.07251588 0.00468886] [0.00368813 0.98887899 0.00376892 0.00366397] [0.00502992 0.00530702 0.83905815 0.15060491] [0.00928782 0.9717597  0.00996268 0.0089898 ] [0.98160245 0.00609432 0.00623438 0.00606885]


The values indicate the probability of the document belonging to each topic. For example, the first  document has a very high probability of belonging to the first topic, and very low probability for the other topics. 

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (30).ipynb
    
</div>

In [ ]:
first_five_topic_dist = document_topics[:5]
print(first_five_topic_dist)


# for i in range(5):
#     print("document", i, "has the probability of ",  , " : ",  df.iloc[i]['text_pp'])

for i, doc_dist in enumerate(first_five_topic_dist):
    print(f"Document {i} has the highest probability to be about topic: {doc_dist.argmax()}, Probability: {np.max(doc_dist)}")
print("\n")
print("Model 0:")
print(df.iloc[0]['text_pp'])
print("Model 0 topic assignemnt:")
print(first_five_topic_dist[0])


[[0.00236199 0.19409508 0.32895831 0.46513699 0.00236188 0.00236292
  0.00235909 0.00236374]
 [0.41087621 0.00176359 0.00176283 0.0017627  0.57854635 0.00176326
  0.00176258 0.00176248]
 [0.00240802 0.82127972 0.00240777 0.00240637 0.16427395 0.00240863
  0.00240684 0.0024087 ]
 [0.00521022 0.005213   0.00521923 0.00521202 0.46679663 0.50192308
  0.00520896 0.00521686]
 [0.00277972 0.00277967 0.98054155 0.00277882 0.00278016 0.00277925
  0.00277875 0.00278207]]
Document 0 has the highest probability to be about topic: 3, Probability: 0.465136994087788
Document 1 has the highest probability to be about topic: 4, Probability: 0.5785463490408699
Document 2 has the highest probability to be about topic: 1, Probability: 0.8212797238598143
Document 3 has the highest probability to be about topic: 5, Probability: 0.5019230828043948
Document 4 has the highest probability to be about topic: 2, Probability: 0.9805415490795232


Model 0:
know read sence wonder help provide paragraph shall unlawfu

The model seems to be fairly confident about some assignments (document 4 and 2) and somewhat uncertain about the rest (with probability ranging from .46 to .57). In all the cases, however, the model is choosing the best topic assignment for the document. For instance, the model is the least certain about model 0 (printed above). It's 47% certain that the topic for this document is topic 3 or Firearms Control, which seems to be accurate.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (18).ipynb
    
</div>

In [ ]:
doc_topic_matrix = lda_model.transform(X)
for i in range(5):
    print(f"Document {i + 1}:")
    print(doc_topic_matrix[i])
    print()


- Document 1: The highest probability (0.956) that it belongs to Topic 6, which is about the Israel Crisis.
- Document 2: Has the probability (0.721) of belonging to Topic 4, which is about Computer Graphics.
- Document 3: The highest probability (0.403) that it belongs to Topic 6, which is about the Israel Crisis. Though this is a bit strange, and the probability is much lower than needed. Hence I think this document is probably NOT related to that.
- Document 4: The highest probability that it belongs to Topic 4, which is also about Computer Graphics.
- Document 5: The highest probability that it belongs to Topic 10, which is about the country of Turkey.


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (14).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f'Document {i} is made up of the weights: {document_topics_weights[i]}')
    print(f'Document {i} is most likely assigned to topic {np.argmax(document_topics_weights[i])}\n')

Document 0 is made up of the weights: [0.19093249 0.79916813 0.00498515 0.00491423]
Document 0 is most likely assigned to topic 1

Document 1 is made up of the weights: [0.00370779 0.13720637 0.85535288 0.00373296]
Document 1 is most likely assigned to topic 2

Document 2 is made up of the weights: [0.00495358 0.67228184 0.31804456 0.00472002]
Document 2 is most likely assigned to topic 1

Document 3 is made up of the weights: [0.14568616 0.00757973 0.74118767 0.10554645]
Document 3 is most likely assigned to topic 2

Document 4 is made up of the weights: [0.98208821 0.0060162  0.0059796  0.00591599]
Document 4 is most likely assigned to topic 0



The document topic assignment of the model seems to be pretty accurate based on the first 5 sentences.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (4).ipynb
    
</div>

In [ ]:
print(document_topics[0])
print(document_topics[1])
print(document_topics[2])
print(document_topics[3])
print(document_topics[4])

[0.00231614 0.56182299 0.00231675 0.0023184  0.00231566 0.0023181
 0.4242764  0.00231556]
[0.00176259 0.00176298 0.00176341 0.05465663 0.00176216 0.00176439
 0.12865503 0.80787281]
[0.00227312 0.00227643 0.00227456 0.9840704  0.0022752  0.00227783
 0.0022778  0.00227466]
[0.00431086 0.00431894 0.00431333 0.61939602 0.00431239 0.00431645
 0.06137983 0.29765217]
[0.00291089 0.0029114  0.0029118  0.00291009 0.26750605 0.00291291
 0.71502586 0.00291099]


Doc 1 was classified as guns and politics. Doc 2 was computer images. Doc 3 and doc 4 were computer interface. Doc 5 was christianity. Overall, these 5 documents were matched pretty closely with their original topics.

<div class="alert alert-warning">

hw7_sol (8).ipynb
    
</div>

In [ ]:
lda.transform(X[0]),lda.transform(X[1]),lda.transform(X[2]),lda.transform(X[3]),lda.transform(X[4])

(array([[0.00368406, 0.00367576, 0.98530902, 0.00365847, 0.00367268]]),
 array([[0.03902386, 0.73060634, 0.00290067, 0.22459329, 0.00287583]]),
 array([[0.00342   , 0.35883096, 0.63088703, 0.00342372, 0.0034383 ]]),
 array([[0.00675827, 0.00677546, 0.00682016, 0.00669198, 0.97295413]]),
 array([[0.92053179, 0.00449157, 0.06602349, 0.00448094, 0.0044722 ]]))

In [ ]:
df["text"][0]
#For document 1, the model assigns this to 'politics or geopolitics', and this seems fairly accurate. 

'You know, I was reading 18 U.S.C. 922 and something just did not make \nsence and I was wondering if someone could help me out.\n\nSay U.S.C. 922 :\n\n(1) Except as provided in paragraph (2), it shall be unlawful for\nany person to transfer or possess a machinegun.\n\n Well I got to looking in my law dictionary and I found that a "person" \nmight also be an artificial entity that is created by government \nand has no rights under the federal constitution. So, what I \ndon\'t understand is how a statute like 922 can be enforced on \nan individual. So someone tell me how my government can tell\nme what I can or cannot possess. Just passing a law \ndoes not make it LAW. Everyone knows that laws are constitional\nuntil it goes to court. So, has it ever gone to court, not\njust your run of the mill "Ok I had it I am guilty, put me in jail"\n\nHas anyone ever claimed that they had a right to possess and was told\nby the Supreme Court that they didn\'t have that right?\n\n\n'

In [ ]:
df["text"][1]
#For document 2, the model assigns it more to 'Software' than any other topic, with 45%, and it seems to be fairly accurate, even if it isn't exactly software, it is closer to that topic than any other topic.


"\n\n\nIt's not a bad question: I don't have any refs that list this algorithm\neither. But thinking about it a bit, it shouldn't be too hard.\n\n1) Take three of the points and find the plane they define as well as\nthe circle that they lie on (you say you have this algorithm already)\n\n2) Find the center  of this circle. The line passing through this center\nperpendicular to the plane of the three points passes through the center of\nthe sphere.\n\n3) Repeat with the unused point and two of the original points. This\ngives you two different lines that both pass through the sphere's\norigin. Their interection is the center of the sphere.\n\n4) the radius is easy to compute, it's just the distance from the center to\nany of the original points.\n\nI'll leave the math to you, but this is a workable algorithm. :-)\n\n\nAn alternate method would be to take pairs of points: the plane formed\nby the perpendicular bisector of each line segment pair also contains the\ncenter of the sphere. T

In [ ]:
df["text"][2]
#For this, the model is very close between 'Religion' and 'Software'. The model should guess 'Software' as that is what the document is closest to,
# but even though it does guess 'software' with 48.6%, it guesses 'religion' with 50%. So the model could do better.


"\nActuallay I don't, but on the other hand I don't support the idea of having\none newsgroup for every aspect of graphics programming as proposed by Brian,\nin his reply to my original posting.\nI would suggest a looser structure more like a comp.graphics.programmer,\ncomp.graphics.hw_and_sw\nThe reason for making as few groups as possible is for the same reason you\nsay we shouldn't spilt up, not to get to few postings every day.\nI takes to much time to browse through all postings just to find two or \nthree I'm interested in.\n\nI understand and agree when you say you want all aspects of graphics in one\nmeeting. I agree to some extension. I see news as a forum to exchange ideas,\nhelp others or to be helped. I think this is difficult to achive if there\nare so many different things in one meeting.\n\nGood evening netters|-)"

In [ ]:
df["text"][3]
#It confidently guesses 'office technology' and it is correct.

"The following problem is really bugging me,\nand I would appreciate any help.\n\nI create two windows:\n\nw1 (child to root) with event_mask = ButtonPressMask|KeyPressMask;\nw2 (child to w1) with do_not_propagate_mask = ButtonPressMask|KeyPressMask;\n\n\nKeypress events in w2 are discarded, but ButtonPress events fall through\nto w1, with subwindow set to w2.\n\nFYI, I'm using xnews/olvwm.\n\nAm I doing something fundamentally wrong here?"

In [ ]:
df["text"][4]
#It confidently guesses 'politics or geopolitics' and is correct here.

"\n\n  This is the latest from UPI \n\n     Foreign Ministry spokesman Ferhat Ataman told journalists Turkey was\n     closing its air space to all flights to and from Armenia and would\n     prevent humanitarian aid from reaching the republic overland across\n     Turkish territory.\n\n  \n   Historically even the most uncivilized of peoples have exhibited \n   signs of compassion by allowing humanitarian aid to reach civilian\n   populations. Even the Nazis did this much.\n\n   It seems as though from now on Turkey will publicly pronounce \n   themselves 'hypocrites' should they choose to continue their\n   condemnation of the Serbians.\n\n\n\n--"

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (10).ipynb
    
</div>

In [ ]:
for i in range(5):
    document_topics = lda.transform(X[i])
    print(f"Document {i+1} Topic Assignment: {document_topics.argmax()}")

Document 1 Topic Assignment: 1
Document 2 Topic Assignment: 2
Document 3 Topic Assignment: 2
Document 4 Topic Assignment: 0
Document 5 Topic Assignment: 4


Document 1 is likely some sort of religious/thoughful piece of writing. Documents 2, 3 are likely about computer file formats or some sort of computer manual related text. Document 4 may also be some sort of computer manual related text but less emphasis on file types. Document 5 is likely about judaism. 


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (49).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(df.iloc[i]["text_pp"])
    print(document_topics[i])
    print("")

know read u.s.c. sence wonder help u.s.c. provide paragraph unlawful person transfer possess machinegun   get look law dictionary find person artificial entity create government right federal constitution understand statute enforce individual tell government tell possess pass law law know law constitional go court go court run mill guilty jail claim right possess tell supreme court right 
[0.00227465 0.00227449 0.00227711 0.00227704 0.03264882 0.62783092
 0.32814355 0.00227341]

 bad question ref list algorithm think bit hard point find plane define circle lie algorithm find center circle line pass center perpendicular plane point pass center sphere repeat unused point original point give different line pass sphere origin interection center sphere radius easy compute distance center original point leave math workable algorithm  alternate method pair point plane form perpendicular bisector line segment pair contain center sphere pair form plane intersect point easy implement
[0.69250835

Overall, the document assignments were fairly accurate in picking the general topics. However, as each topic had a corresponding similar topic(Ex. Religion with Christianity and Atheism), the model wasn't completely confident in picking one of the two. For example, on the fourth example that appears to be about computers, it gets stuck between two options which makes sense as there are two topics related to computers. 

<div class="alert alert-warning">

hw7_sol (34).ipynb
    
</div>

In [ ]:
# Get the document-topic distribution for the entire dataset
doc_topic_dist = lda.transform(X)

# Show the document-topic assignment for the first five documents
first_five_docs = doc_topic_dist[:5]

# Display the distribution
print("Document-Topic Distribution for the First Five Documents:")
for i, doc_distribution in enumerate(first_five_docs):
    print(f"Document {i+1}: {doc_distribution}")

# Comment on the document topic assignment
for i, topic_mixture in enumerate(first_five_docs):
    dominant_topic = np.argmax(topic_mixture)
    print(f"Document {i+1} is most associated with Topic {dominant_topic}, with a proportion of {topic_mixture[dominant_topic]:.4f}.")


Comments on Document Topic Assignment:

Document 1: This document has a strong association with Topic 5, where approximately 75.8% of the document's content pertains to this topic. This suggests that the content of Document 1 is relatively focused on a single theme that Topic 5 represents.

Document 2: Similarly to Document 1, Document 2 is predominantly associated with Topic 3, with a proportion of about 79.6%. This again indicates a strong focus on a single topic within this document.

Document 3: This document has its content spread across multiple topics, but the dominant topic is Topic 0 with a 36.5% share. The presence of other topics with significant proportions could suggest that the document touches upon multiple themes.

Document 4: Shows a very strong association with Topic 1 (87.1%). This indicates that the document is very likely to be discussing the subject matter of Topic 1 almost exclusively.

Document 5: This document is most associated with Topic 2, but the proportion (47.1%) is not as overwhelming as in Documents 1, 2, and 4. This suggests that while Topic 2 is a significant theme, other topics also play a notable role in the content of this document.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (45).ipynb
    
</div>

In [ ]:
df.head(5)

text  target  \
0  You know, I was reading 18 U.S.C. 922 and some...       6   
1  \n\n\nIt's not a bad question: I don't have an...       1   
2  \nActuallay I don't, but on the other hand I d...       1   
3  The following problem is really bugging me,\na...       2   
4  \n\n  This is the latest from UPI \n\n     For...       7   

             target_name                                            text_pp  
0     talk.politics.guns  know read sence wonder help provide paragraph ...  
1          comp.graphics  bad question ref list algorithm think bit hard...  
2          comp.graphics  actuallay hand support idea have newsgroup asp...  
3         comp.windows.x  follow problem bug appreciate help create wind...  
4  talk.politics.mideast  late upi foreign ministry spokesman ferhat ata...

In [ ]:
df.iloc[0]["text"]

'You know, I was reading 18 U.S.C. 922 and something just did not make \nsence and I was wondering if someone could help me out.\n\nSay U.S.C. 922 :\n\n(1) Except as provided in paragraph (2), it shall be unlawful for\nany person to transfer or possess a machinegun.\n\n Well I got to looking in my law dictionary and I found that a "person" \nmight also be an artificial entity that is created by government \nand has no rights under the federal constitution. So, what I \ndon\'t understand is how a statute like 922 can be enforced on \nan individual. So someone tell me how my government can tell\nme what I can or cannot possess. Just passing a law \ndoes not make it LAW. Everyone knows that laws are constitional\nuntil it goes to court. So, has it ever gone to court, not\njust your run of the mill "Ok I had it I am guilty, put me in jail"\n\nHas anyone ever claimed that they had a right to possess and was told\nby the Supreme Court that they didn\'t have that right?\n\n\n'

In [ ]:
for i in range(5):
    print("The document topic assignment for document {} is: {}".format(i, document_topics[i]))
    print("The majority topic for document {} is topic {}".format(i, np.argmax(document_topics[i])))

The document topic assignment for document 0 is: [0.00503981 0.00493475 0.00508405 0.98494138]
The majority topic for document 0 is topic 3
The document topic assignment for document 1 is: [0.84326901 0.00366354 0.14929961 0.00376784]
The majority topic for document 1 is topic 0
The document topic assignment for document 2 is: [0.00519156 0.00498815 0.3987506  0.59106968]
The majority topic for document 2 is topic 3
The document topic assignment for document 3 is: [0.00746733 0.00723032 0.9768585  0.00844385]
The majority topic for document 3 is topic 2
The document topic assignment for document 4 is: [0.16277004 0.00586893 0.00595576 0.82540527]
The majority topic for document 4 is topic 3


The document topic assignment is not that accurate because for example document 2 which is about computer graphics is being assigned to religion which is not accurate. There are too many words in each document which may make it difficult to accurately determine the topic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (38).ipynb
    
</div>

In [ ]:
document_topics = lda_model.transform(doc_term_matrix)
print(document_topics[:5])

[[0.00231616 0.9837875  0.00231683 0.00231511 0.00231525 0.00231707
  0.00231688 0.00231521]
 [0.00176304 0.3774834  0.00176267 0.00176169 0.00176203 0.00176261
  0.61194359 0.00176098]
 [0.00227449 0.64304147 0.00227515 0.00227301 0.00227515 0.00227499
  0.34331193 0.0022738 ]
 [0.00431266 0.12266272 0.00431812 0.00431444 0.00431147 0.00431633
  0.44986377 0.40590049]
 [0.00290994 0.00291422 0.00291041 0.00290714 0.00290885 0.97963205
  0.00291009 0.0029073 ]]


1. **Document 1** is almost completely about one topic (about 98% on topic 2). This means it's really focused on just one subject.
2. **Document 2** has a mix but leans heavily towards topic 7 (about 61%), with some parts about topic 2 (about 38%). It's like a cake with two main flavors, but one is stronger.
3. **Document 3** is also a mix between topics 2 and 7, kind of like Document 2, showing it talks about two main things.
4. **Document 4** is more evenly spread out over a few topics, especially topics 7 and 8. This one's like a blend of several flavors, not just one or two.
5. **Document 5** is overwhelmingly about topic 6 (almost 98%). It's super focused, just like Document 1, but on a different topic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (26).ipynb
    
</div>

In [ ]:
doc_topic_dist = lda.transform(dtm)

for doc_idx, doc_dist in enumerate(doc_topic_dist[:5]):
    print("Document {}:".format(doc_idx))
    for topic_idx, topic_prob in enumerate(doc_dist):
        print("  Topic {}: {:.4f}".format(topic_idx, topic_prob))
    print()

Document 0:
  Topic 0: 0.0426
  Topic 1: 0.0025
  Topic 2: 0.0025
  Topic 3: 0.0260
  Topic 4: 0.1616
  Topic 5: 0.0025
  Topic 6: 0.0025
  Topic 7: 0.7599

Document 1:
  Topic 0: 0.0018
  Topic 1: 0.0018
  Topic 2: 0.0799
  Topic 3: 0.0594
  Topic 4: 0.0018
  Topic 5: 0.0018
  Topic 6: 0.8003
  Topic 7: 0.0532

Document 2:
  Topic 0: 0.0026
  Topic 1: 0.0026
  Topic 2: 0.0026
  Topic 3: 0.0026
  Topic 4: 0.0026
  Topic 5: 0.3509
  Topic 6: 0.0026
  Topic 7: 0.6338

Document 3:
  Topic 0: 0.0057
  Topic 1: 0.0057
  Topic 2: 0.9602
  Topic 3: 0.0057
  Topic 4: 0.0057
  Topic 5: 0.0057
  Topic 6: 0.0057
  Topic 7: 0.0057

Document 4:
  Topic 0: 0.9398
  Topic 1: 0.0031
  Topic 2: 0.0031
  Topic 3: 0.0031
  Topic 4: 0.0031
  Topic 5: 0.0031
  Topic 6: 0.0419
  Topic 7: 0.0031



- Document 0: This document shows a strong link to Topic 7.
- Document 1: This document seems to be almost in Topics 6.
- Document 2: This document seems to be almost split by Topics 5 and 7, suggesting it covers two subjects, but is less concentrated compared to Document 7.
- Document 3: This document seems to be almost in topics 2 as well.
- Document 4: Here we see a strong link towards Topic 0.

<div class="alert alert-warning">

hw7_sol.ipynb
    
</div>

In [ ]:
res = document_topics[:5]
print("first 5 assignments:")
res

first 5 assignments:


array([[0.00515185, 0.00504452, 0.00520811, 0.98459553],
       [0.88395404, 0.00385295, 0.00398239, 0.10821063],
       [0.41628715, 0.08497786, 0.00451799, 0.494217  ],
       [0.97318636, 0.0085905 , 0.00880172, 0.00942143],
       [0.00629393, 0.00616323, 0.54640409, 0.44113875]])

In [ ]:
df["text"][:5]

0    You know, I was reading 18 U.S.C. 922 and some...
1    \n\n\nIt's not a bad question: I don't have an...
2    \nActuallay I don't, but on the other hand I d...
3    The following problem is really bugging me,\na...
4    \n\n  This is the latest from UPI \n\n     For...
Name: text, dtype: object

The model did a fairly good job, the 4 out of the first 5 were assigned to the corrct topic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (2).ipynb
    
</div>

In [ ]:
doc_topic_dist = lda_model.transform(dtm[:5])

# Display the document topic assignment
for doc_idx, dist in enumerate(doc_topic_dist):
    print(f"Document #{doc_idx + 1} Topic Distribution:")
    for topic_idx, topic_prob in enumerate(dist):
        print(f"  Topic {topic_idx}: {topic_prob:.4f}")
    print()

Document #1:
- Dominantly about Topic 2: This document is overwhelmingly associated with Topic 2 (98.49%). This indicates that the content of Document #1 is highly focused on the theme represented by Topic 2. The document likely contains specific language or terms that are characteristic of this topic.

Document #2:
- Mixed Topics with a Preference for Topic 0: Document #2 shows a notable distribution across Topics 0, 4, and 5, with Topic 0 being the most dominant. This suggests that the document covers multiple themes but leans towards the subject matter of Topic 0.

Document #3:
- Balance between Topics 0, 2, and 7: This document is somewhat evenly split among Topics 0, 2, and 7, with Topic 7 being slightly more dominant. It suggests a document touching upon various subjects, making it a multifaceted discussion.

Document #4:
- Predominantly Topics 0 and 2: With more than half of the document aligned with Topic 0 and a significant portion associated with Topic 2, Document #4 seems to primarily discuss these two themes. It shows a clear inclination towards the subjects of Topics 0 and 2.

Document #5:
- Mainly Topics 2 and 5: This document is mostly about Topic 2, with a notable mention of Topic 5. It seems to focus on the area represented by Topic 2 while also covering some aspects of Topic 5.


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (12).ipynb
    
</div>

_Points:_ 5

In [ ]:
document_topics

array([[4.32647905e-03, 4.51260822e-03, 9.86813686e-01, 4.34722643e-03],
       [3.58147661e-03, 3.48588922e-03, 5.45224686e-01, 4.47707948e-01],
       [4.44541747e-03, 4.40141943e-03, 6.94225979e-01, 2.96927184e-01],
       ...,
       [5.21860351e-03, 2.56609152e-01, 7.32936621e-01, 5.23562416e-03],
       [6.92985279e-03, 6.48792974e-03, 6.59336302e-03, 9.79988854e-01],
       [7.91038698e-04, 4.39497665e-01, 5.58911465e-01, 7.99830672e-04]])

In [ ]:
for i in range(5):
    print("Document", i + 1, ":\n", df.iloc[i]['text_pp'])
    print("Document", i + 1, "topic assignment:\n", document_topics[i])
    print("\n")

Document 1 :
 know read u.s.c. 922 sence wonder help u.s.c. 922 provide paragraph shall unlawful person transfer possess machinegun get look law dictionary find person artificial entity create government right federal constitution understand statute 922 enforce individual tell government tell possess pass law law know law constitional go court go court run mill guilty jail claim right possess tell supreme court right
Document 1 topic assignment:
 [0.00432648 0.00451261 0.98681369 0.00434723]


Document 2 :
 bad question ref list algorithm think bit hard point find plane define circle lie algorithm find center circle line pass center perpendicular plane point pass center sphere repeat unused point original point give different line pass sphere origin interection center sphere radius easy compute distance center original point leave math workable algorithm alternate method pair point plane form perpendicular bisector line segment pair contain center sphere pair form plane intersect point

1. Document 1 seems to be assigned to the right topic (religion).
2. Document 2 doesn't seem to align heavily with any of the topics, but the closest assignment is for religion. However, we expected it to be more closely aligned with computers (which the model does have it as a close second).
3. Document 3 should've been assigned to topic 3 (computers), but the model incorrectly assigned it to religion.
4. Document 4 seems to be assigned to the right topic (computers).
5. Document 5 seems to be assigned to the right topic (middle eastern politics).

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (24).ipynb
    
</div>

In [ ]:
res = document_topics[:5]
print("first 5 assignments:")
res

In [ ]:
df["text"][:5]

The model did a fairly good job, the 4 out of the first 5 were assigned to the corrct topic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (47).ipynb
    
</div>

In [ ]:
print(df.iloc[0]["text"])
print(document_topics[0])

You know, I was reading 18 U.S.C. 922 and something just did not make  sence and I was wondering if someone could help me out.  Say U.S.C. 922 :  (1) Except as provided in paragraph (2), it shall be unlawful for any person to transfer or possess a machinegun.   Well I got to looking in my law dictionary and I found that a "person"  might also be an artificial entity that is created by government  and has no rights under the federal constitution. So, what I  don't understand is how a statute like 922 can be enforced on  an individual. So someone tell me how my government can tell me what I can or cannot possess. Just passing a law  does not make it LAW. Everyone knows that laws are constitional until it goes to court. So, has it ever gone to court, not just your run of the mill "Ok I had it I am guilty, put me in jail"  Has anyone ever claimed that they had a right to possess and was told by the Supreme Court that they didn't have that right?   
[0.0047665  0.73705124 0.00474329 0.25343

In [ ]:
print(df.iloc[1]["text"])
print(document_topics[1])

   It's not a bad question: I don't have any refs that list this algorithm either. But thinking about it a bit, it shouldn't be too hard.  1) Take three of the points and find the plane they define as well as the circle that they lie on (you say you have this algorithm already)  2) Find the center  of this circle. The line passing through this center perpendicular to the plane of the three points passes through the center of the sphere.  3) Repeat with the unused point and two of the original points. This gives you two different lines that both pass through the sphere's origin. Their interection is the center of the sphere.  4) the radius is easy to compute, it's just the distance from the center to any of the original points.  I'll leave the math to you, but this is a workable algorithm. :-)   An alternate method would be to take pairs of points: the plane formed by the perpendicular bisector of each line segment pair also contains the center of the sphere. Three pairs will form three

In [ ]:
print(df.iloc[2]["text"])
print(document_topics[2])

 Actuallay I don't, but on the other hand I don't support the idea of having one newsgroup for every aspect of graphics programming as proposed by Brian, in his reply to my original posting. I would suggest a looser structure more like a comp.graphics.programmer, comp.graphics.hw_and_sw The reason for making as few groups as possible is for the same reason you say we shouldn't spilt up, not to get to few postings every day. I takes to much time to browse through all postings just to find two or  three I'm interested in.  I understand and agree when you say you want all aspects of graphics in one meeting. I agree to some extension. I see news as a forum to exchange ideas, help others or to be helped. I think this is difficult to achive if there are so many different things in one meeting.  Good evening netters -)
[0.40386252 0.58689191 0.00458155 0.00466402]


In [ ]:
print(df.iloc[3]["text"])
print(document_topics[3])

The following problem is really bugging me, and I would appreciate any help.  I create two windows:  w1 (child to root) with event_mask   ButtonPressMask KeyPressMask; w2 (child to w1) with do_not_propagate_mask   ButtonPressMask KeyPressMask;   Keypress events in w2 are discarded, but ButtonPress events fall through to w1, with subwindow set to w2.  FYI, I'm using xnews olvwm.  Am I doing something fundamentally wrong here?
[0.89574104 0.08604298 0.00906574 0.00915025]


In [ ]:
print(df.iloc[4]["text"])
print(document_topics[4])

    This is the latest from UPI        Foreign Ministry spokesman Ferhat Ataman told journalists Turkey was      closing its air space to all flights to and from Armenia and would      prevent humanitarian aid from reaching the republic overland across      Turkish territory.        Historically even the most uncivilized of peoples have exhibited     signs of compassion by allowing humanitarian aid to reach civilian    populations. Even the Nazis did this much.     It seems as though from now on Turkey will publicly pronounce     themselves 'hypocrites' should they choose to continue their    condemnation of the Serbians.    --
[0.00602719 0.4190131  0.00582867 0.56913104]


We can see that each text would have a higher probability of one kind of topic and this would help us know the kind of the text and do the split of topic.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (36).ipynb
    
</div>

In [ ]:
document_topics_five = lda.fit_transform(X[0:5])
sorting = np.argsort(lda.components_, axis=1)[:, ::-1]
feature_names = np.array(vec.get_feature_names_out())

In [ ]:
import mglearn

mglearn.tools.print_topics(
    topics=range(5),
    feature_names=feature_names,
    sorting=sorting,
    topics_per_chunk=5,
    n_words=10,
)

topic 0       topic 1       topic 2       topic 3       topic 4       
--------      --------      --------      --------      --------      
create        create        w2            law           point         
help          help          w1            help          center        
tell          tell          keypressmask  tell          plane         
different     different     child         posting       sphere        
think         think         event         possess       pass          
pass          pass          buttonpressmaskright         pair          
original      original      keypress      court         line          
subwindow     subwindow     xnew          idea          algorithm     
problem       problem       bug           reason        original      
fundamentally fundamentally fall          aspect        circle        




<span style="color:blue"> 

The document topic assignment for the five topics is very different compared to the five topics I identified above based on the whole data frame. The first 2 topics are practically the same, topic 2 is a mix of different words, topic 3 may be more related to weapons and the government, and topic 4 is related to geometry it seems. Therefore, observing the 5 topics based on the first five lines of `df` does not provide a good overview of the general topics found in `df`. 

</span> 

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (28).ipynb
    
</div>

In [ ]:
print(document_topics[0]) 
# it is very confident (0.965) to be topic 0 (about politics), and it's correct when compared to original target name
print(document_topics[1]) 
# it is likely (0.483) to be topic 1 (about computer),and it's correct when compared to original target name
print(document_topics[2]) 
# it is very likely (0.729) to be topic 4 (about religion), but actually the orginal target name is in the computer category, the model predicted 0.26 (second possible) for topic 1 (computer)
print(document_topics[3])
# it is very likely (0.668) to be topic 1 (about computer),and it's correct when compared to original target name
print(document_topics[4]) 
# it is very confident (0.965) to be topic 0 (about politics), and it's correct when compared to original target name

[0.96523291 0.0049314  0.02491887 0.00491681]
[0.21425712 0.48351845 0.00371562 0.29850881]
[0.00509117 0.26003577 0.00498279 0.72989028]
[0.0089665  0.66815072 0.0086817  0.31420109]
[0.98221024 0.0059695  0.00590799 0.00591228]


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (17).ipynb
    
</div>

In [ ]:
pd.set_option('display.max_colwidth', 0)

temp=pd.DataFrame(document_topics[0:4],columns=["topic_0", "topic_1", "topic_2", "topic_3","topic_4","topic_5"])
temp["text"] = df[0:4]["text_pp"]
temp

topic_0   topic_1   topic_2   topic_3   topic_4   topic_5  \
0  0.003369  0.003361  0.003357  0.003346  0.003350  0.983217   
1  0.002332  0.503667  0.002324  0.002332  0.002326  0.487019   
2  0.002755  0.691161  0.002739  0.002758  0.002743  0.297842   
3  0.812753  0.005251  0.005216  0.166328  0.005218  0.005236   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            text  
0  [know, read, u.s.c., sence, wonder, help, u.s.c., provide, paragraph, shall, unlawful, forany, person, transfer, possess, machinegun, look, dictionary, find, person, artificial, entity, create, government, right, federal, constitution, understand, statute, like, enforce, individual, tell, government, tellme, possess, pass, know, law, constitionaluntil, go, court, go, court, notjust, mill, guilty, jail"has, claim, right, possess, toldby, supreme, court, right]                                                                                                                                                                                
1  [question, refs, list, algorithmeither, thinking, hard.1, point, find, plane, define, asthe, circle, algorithm, already)2, find, center, circle, line, pass, centerperpendicular, plane, point, pass, center, ofthe, sphere.3, repeat, unused, point, original, point, thisgive, different, line, pass, sphere'sorigin, interection, center, sphere.4, radius, easy, compute, distance, center, toany, original, point, i'll, leave, math, workable, algorithm, -)an, alternate, method, pair, point, plane, formedby, perpendicular, bisector, line, segment, pair, contain, thecenter, sphere, pair, form, plane, intersectingat, point, easier, implement]  
2  [actuallay, hand, support, idea, havingone, newsgroup, aspect, graphic, programming, propose, brian, reply, original, posting, suggest, loose, structure, like, comp.graphics.programmer, comp.graphics.hw_and_swthe, reason, make, group, possible, reason, yousay, spill, posting, take, time, browse, posting, find, interested, understand, agree, want, aspect, graphic, onemeeting, agree, extension, news, forum, exchange, idea, help, help, think, difficult, achive, thereare, different, thing, meeting, good, evening, netters|-]                                                                                                                  
3  [follow, problem, bug, appreciate, help, create, window, child, root, event_mask, buttonpressmask|keypressmask;w2, child, do_not_propagate_mask, buttonpressmask|keypressmask;keypress, event, discard, buttonpress, event, fall, throughto, subwindow, w2.fyi, xnew, olvwm, fundamentally, wrong]

While some documents are assigned with high confidence to a particular topic (ex. the first document is assigned 98% to topic 5), other documents are much more ambiguous and share almost equal belonging to multiple groups (ex. the second document which is almost evenly divided between topic_1 and topic_5). Low confidence assignments tend to be associated with texts that make excessive use of ambiguous and technical language (not very frequent in other texts).

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (50).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(df['text_pp'].iloc[i])
    print(doc_topics[i])
    print("\n")

know read 922 sence wonder help 922 provide paragraph shall unlawful person transfer possess machinegun get look law dictionary find person artificial entity create government right federal constitution understand statute 922 enforce individual tell government tell possess pass law law know law constitional go court go court run mill guilty jail claim right possess tell supreme court right
[0.72263598 0.2661032  0.00377531 0.00374306 0.00374245]


bad question ref list algorithm think bit hard point find plane define circle lie algorithm find center circle line pass center perpendicular plane point pass center sphere repeat unused point original point give different line pass sphere origin interection center sphere radius easy compute distance center original point leave math workable algorithm alternate method pair point plane form perpendicular bisector line segment pair contain center sphere pair form plane intersect point easy implement
[0.00287243 0.00288958 0.98850707 0.00285242 

The topic assignment of this model seems generally correct. The model is very confident that the second example and fourth example is assigned to the 'Computers' topic, and the fifth example is assigned to the 'Middle East' topic, which we would agree with. For the first example, the model is mainly confident it belongs to the 'Gun Crime' topic, but is also slightly considering the 'Christian Religion' topic. We think it shouldn't be considering the 'Christian Religion' topic as the words in example 1 are all mainly related to crime. However, the model is mainly confident that the third example belongs in the 'Christian Religion' topic, but also slightly considering the 'Computers' topic. We think the model is wrong with this example as the words in example 3 are mainly related to programming and corporate workspaces, which aren't really related to religion at all. 

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (33).ipynb
    
</div>

In [ ]:
df.head()

text  target  \
0  You know, I was reading 18 U.S.C. 922 and some...       6   
1  \n\n\nIt's not a bad question: I don't have an...       1   
2  \nActuallay I don't, but on the other hand I d...       1   
3  The following problem is really bugging me,\na...       2   
4  \n\n  This is the latest from UPI \n\n     For...       7   

             target_name                                            text_pp  
0     talk.politics.guns  u.s.c. 922 sence u.s.c. 922 paragraph shall un...  
1          comp.graphics  bad question ref algorithm bit hard point plan...  
2          comp.graphics  actuallay hand idea newsgroup aspect graphic p...  
3         comp.windows.x  problem help window child event_mask buttonpre...  
4  talk.politics.mideast  late upi foreign ministry spokesman ferhat ata...

In [ ]:
pd.DataFrame(document_topics).head()

0         1         2         3         4         5         6
0  0.004212  0.004212  0.004247  0.974691  0.004223  0.004208  0.004206
1  0.067709  0.002555  0.702251  0.002556  0.219809  0.002558  0.002562
2  0.003880  0.003869  0.003873  0.003869  0.577855  0.003870  0.402783
3  0.006826  0.196190  0.165033  0.283086  0.006828  0.006814  0.335222
4  0.004340  0.805683  0.004332  0.004352  0.004345  0.172613  0.004336

1. the first document is mainly related to topic 3 (0.975) (being less than 1% of each of the other topics), which is about weapon(gun) politics, which matches the df.
2. second document is made up of 70% topic 2, 22% topic 4, and 7% topic 0 (and less than 1% of each of the other topics). Topic 2 is about baseball which does not match too well since it's labeled as computer graphics topic.
3. third document is made up of 58% topic 4, 40% topic 6, and < 1% the others, which is slightly on point when compared to the labeled topic since it's computer graphics, but topic 4 is christianity while topic 6 is about computers.
4. the fourth document is made up of 34% topic 6, 28% topic 3, and 20% topic 1, and 16% topic 2 (and < 1% of the other topics). It's labelled as a computer topic, which is topic 6 and is the largest proportion of the topics.
5. fifth document is 81% topic 1, 17% topic 5, and less than 1% of each of the other topics. It's labeled as middle east politics talk, which is topic 1, which is a good match making the majority of the topic assignment.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (42).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f"Document {i}:")
    print(document_topics[i])

Document 0:
[0.00231614 0.56182299 0.00231675 0.0023184  0.00231566 0.0023181
 0.4242764  0.00231556]
Document 1:
[0.00176259 0.00176298 0.00176341 0.05465663 0.00176216 0.00176439
 0.12865503 0.80787281]
Document 2:
[0.00227312 0.00227643 0.00227456 0.9840704  0.0022752  0.00227783
 0.0022778  0.00227466]
Document 3:
[0.00431086 0.00431894 0.00431333 0.61939602 0.00431239 0.00431645
 0.06137983 0.29765217]
Document 4:
[0.00291089 0.0029114  0.0029118  0.00291009 0.26750605 0.00291291
 0.71502586 0.00291099]


2.
- Document 0: made up of 0.23% topic 0, 56% topic 1, 0.23% topic 2, 0.23% topic 3, 0.23% topic 4, 0.23% topic 5, 42% topic 6, and 0.23% topic 7
- Document 1: made up of 0.17% topics 0, 1, 2, 4, and 5, 5.5% topic 3, 12.9% topic 6, and 80.8% topic 7
- Document 2: made up of 0.22% topics 0, 1, 2, 4, 5, 6, 7, and 98.4% topic 3
- Document 3: made up of 0.43% topics 0, 1, 2, 4, and 5, 61.9% topic 3, 6.1% topic 6 and 29.8% topic 7
- Document 4: made up of 0.29% topics 0, 1, 2, 3, 5, and 7, 26.8% topic 4, and 71.5% topic 6

Looking at the documents, there is a common patterns that topics 0, 1, 2, and 5 do not make up the majority of the documents, while topics 3, 4, 6, and 7 vary. It's interesting because topics 3, 4, and 7 share a common theme of computer and software things, but topic 6 seems related to religious actions. 

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (21).ipynb
    
</div>

In [ ]:
#topic association for the first five documents from df
topic_association = topics[:5]
print(topic_association)

[[0.00509842 0.26070902 0.72919543 0.00499713]
 [0.98876845 0.0037893  0.00372346 0.00371879]
 [0.24286817 0.74692191 0.00511761 0.00509231]
 [0.8930821  0.08391634 0.01195835 0.01104321]
 [0.00603852 0.00613337 0.98191356 0.00591455]]


the tipic assigment is quite clear for the documents. Both document 1 and 5 has high probability of being in topic 3 out of the 4 topics. both doc 2 and 4 have a high probability of being in topic 1, while doc 3 have a decently high prob of topic 2. The probability is reasonably high for 1 topic for each document, which shows that the split is quite clear.

<div class="alert alert-warning">

hw7_sol (7).ipynb
    
</div>

In [ ]:
for i in range(5):
    doc_topics = document_topics[i]
    print(f"Document {i}:")
    for topic_idx, topic_prob in enumerate(doc_topics):
        print(f" - Topic {topic_idx}: {topic_prob}")
    print("\n")

Document 0:
 - Topic 0: 0.002179349074471226
 - Topic 1: 0.3844420169555887
 - Topic 2: 0.0021813219950709887
 - Topic 3: 0.6003001590335817
 - Topic 4: 0.002178844740465059
 - Topic 5: 0.0021799520294579285
 - Topic 6: 0.0021793443136947695
 - Topic 7: 0.002179431348997037
 - Topic 8: 0.00217958050867259


Document 1:
 - Topic 0: 0.0015659277659796826
 - Topic 1: 0.17403324368302808
 - Topic 2: 0.0015658366254873027
 - Topic 3: 0.16735046378411494
 - Topic 4: 0.0015654757555075458
 - Topic 5: 0.0015661760768474148
 - Topic 6: 0.0015654799748789782
 - Topic 7: 0.44687449710111865
 - Topic 8: 0.2039128992330373


Document 2:
 - Topic 0: 0.102706490637526
 - Topic 1: 0.604038984991021
 - Topic 2: 0.002021568271752213
 - Topic 3: 0.0020217451346012123
 - Topic 4: 0.002020668714078451
 - Topic 5: 0.0020215041789265844
 - Topic 6: 0.0020212913223235274
 - Topic 7: 0.0020209956586205845
 - Topic 8: 0.28112675109115054


Document 3:
 - Topic 0: 0.24300665941788274
 - Topic 1: 0.00383561006093

Although we did indeed train on 8 categories and created 9 topics, we would then expect to see that our documents line up with its original labelled category and indeed we do see that, so that's pretty good.Each of the five documents clearly has a topic that has dominant weight. Although we must note that Document 1 has less of a dominant weight (only 0.4) which indicates this document may contain content that is easily present or spreaded out to different topics. We note the consistency between topics and its' associating original category and document nonetheless.  

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (5).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f"topic assignment {i}: ", document_topics[i])
    print(f'document {i}  \n' ,df.iloc[i]['text'])

topic assignment 0:  [0.28540936 0.70442289 0.00507633 0.00509143]
document 0  
 You know, I was reading 18 U.S.C. 922 and something just did not make 
sence and I was wondering if someone could help me out.

Say U.S.C. 922 :

(1) Except as provided in paragraph (2), it shall be unlawful for
any person to transfer or possess a machinegun.

 Well I got to looking in my law dictionary and I found that a "person" 
might also be an artificial entity that is created by government 
and has no rights under the federal constitution. So, what I 
don't understand is how a statute like 922 can be enforced on 
an individual. So someone tell me how my government can tell
me what I can or cannot possess. Just passing a law 
does not make it LAW. Everyone knows that laws are constitional
until it goes to court. So, has it ever gone to court, not
just your run of the mill "Ok I had it I am guilty, put me in jail"

Has anyone ever claimed that they had a right to possess and was told
by the Supreme Cou

The topic assignment is mostly correct. It correctly identified politics and gave the second highest assignment value to religion which makes sense because politics and religions are often talked about in a similar manner. When the topic was computers, the topic assignment succeeded and gave the computer topic a considerably higher score except for the third document. It strongly labeled the third document as religion when the topic is computers. Some misinterpretations are to be expected from our model but we are surprised because the document mostly uses computer vocabulary. However, the last and first paragraph could somehow also be misinterpreted as if it was talking about a religious meeting.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (9).ipynb
    
</div>

In [ ]:
topic_df = pd.DataFrame(topics)
topic_df.columns = ['Topic 0', 'Topic 1', 'Topic 2', 'Topic 3', 'Topic 4', 'Topic 5', 'Topic 6', 'Topic 7']
topic_df
first_five = df.head().join(topic_df)
first_five

text  target  \
0  You know, I was reading 18 U.S.C. 922 and some...       6   
1  \n\n\nIt's not a bad question: I don't have an...       1   
2  \nActuallay I don't, but on the other hand I d...       1   
3  The following problem is really bugging me,\na...       2   
4  \n\n  This is the latest from UPI \n\n     For...       7   

             target_name                                            text_pp  \
0     talk.politics.guns  know read u.s.c. sence wonder help u.s.c. prov...   
1          comp.graphics  bad question ref list algorithm think bit hard...   
2          comp.graphics  hand support idea have newsgroup aspect graphi...   
3         comp.windows.x  follow problem bug appreciate help create wind...   
4  talk.politics.mideast  late upi foreign ministry spokesman ferhat ata...   

    Topic 0   Topic 1   Topic 2   Topic 3   Topic 4   Topic 5   Topic 6  \
0  0.002504  0.982479  0.002503  0.002501  0.002501  0.002505  0.002503   
1  0.001788  0.001787  0.001787  0.001787  0.001787  0.001787  0.001789   
2  0.184790  0.395797  0.002503  0.002502  0.002502  0.002504  0.311811   
3  0.006591  0.006588  0.006589  0.006580  0.006580  0.359079  0.601410   
4  0.002780  0.002780  0.980542  0.002779  0.002778  0.002781  0.002780   

    Topic 7  
0  0.002503  
1  0.987488  
2  0.097590  
3  0.006584  
4  0.002779

#### Response:

Our model seems to have done a good job at classifying these! We show the texts of the first five examples below, so that we can interpret them as humans, view their provided labels, and compare the topic assignment.
- 0 matches with Topic 1 at 0.98, which is correct (Guns -> Firearms)
- 1 matches with Topic 7 at 0.99, which is not quite correct (Graphics -> Baseball). Examining the test, some of the terminology used (e.g. "pass", "transfer", "possess") are words that could appear in texts about Baseball, even though these words have slightly different meanings in computer graphics/algorithms, which could explain the confusion.
- 2 matches with Topic 1 the most at 0.40, which is also quite incorrect (Graphics -> Atheism). Looking at the text, it looks like a forum post discussing the organization of the news platform the post was made in, and the "Firearms" category was most associated with news articles, which could explain the assignment. Also, Topic 6 ("Computers", which would be more correct) was the second-closest, with 0.31 of the text matching.
- 3 matches with Topic 6 at 0.60, which is correct (Windows -> Computer)
- 4 matches with Topic 2 at 0.98, which is correct (Middle East -> Middle East)

Overall, the model seems to have learned topics like Guns, Middle East, and Computers quite well. However, it does seem to have struggled with differentiating computer graphics from the rest of the items.

In [ ]:
first_five['text'][0]

'You know, I was reading 18 U.S.C. 922 and something just did not make \nsence and I was wondering if someone could help me out.\n\nSay U.S.C. 922 :\n\n(1) Except as provided in paragraph (2), it shall be unlawful for\nany person to transfer or possess a machinegun.\n\n Well I got to looking in my law dictionary and I found that a "person" \nmight also be an artificial entity that is created by government \nand has no rights under the federal constitution. So, what I \ndon\'t understand is how a statute like 922 can be enforced on \nan individual. So someone tell me how my government can tell\nme what I can or cannot possess. Just passing a law \ndoes not make it LAW. Everyone knows that laws are constitional\nuntil it goes to court. So, has it ever gone to court, not\njust your run of the mill "Ok I had it I am guilty, put me in jail"\n\nHas anyone ever claimed that they had a right to possess and was told\nby the Supreme Court that they didn\'t have that right?\n\n\n'

In [ ]:
first_five['text'][1]

"\n\n\nIt's not a bad question: I don't have any refs that list this algorithm\neither. But thinking about it a bit, it shouldn't be too hard.\n\n1) Take three of the points and find the plane they define as well as\nthe circle that they lie on (you say you have this algorithm already)\n\n2) Find the center  of this circle. The line passing through this center\nperpendicular to the plane of the three points passes through the center of\nthe sphere.\n\n3) Repeat with the unused point and two of the original points. This\ngives you two different lines that both pass through the sphere's\norigin. Their interection is the center of the sphere.\n\n4) the radius is easy to compute, it's just the distance from the center to\nany of the original points.\n\nI'll leave the math to you, but this is a workable algorithm. :-)\n\n\nAn alternate method would be to take pairs of points: the plane formed\nby the perpendicular bisector of each line segment pair also contains the\ncenter of the sphere. T

In [ ]:
first_five['text'][2]

"\nActuallay I don't, but on the other hand I don't support the idea of having\none newsgroup for every aspect of graphics programming as proposed by Brian,\nin his reply to my original posting.\nI would suggest a looser structure more like a comp.graphics.programmer,\ncomp.graphics.hw_and_sw\nThe reason for making as few groups as possible is for the same reason you\nsay we shouldn't spilt up, not to get to few postings every day.\nI takes to much time to browse through all postings just to find two or \nthree I'm interested in.\n\nI understand and agree when you say you want all aspects of graphics in one\nmeeting. I agree to some extension. I see news as a forum to exchange ideas,\nhelp others or to be helped. I think this is difficult to achive if there\nare so many different things in one meeting.\n\nGood evening netters|-)"

In [ ]:
first_five['text'][3]

"The following problem is really bugging me,\nand I would appreciate any help.\n\nI create two windows:\n\nw1 (child to root) with event_mask = ButtonPressMask|KeyPressMask;\nw2 (child to w1) with do_not_propagate_mask = ButtonPressMask|KeyPressMask;\n\n\nKeypress events in w2 are discarded, but ButtonPress events fall through\nto w1, with subwindow set to w2.\n\nFYI, I'm using xnews/olvwm.\n\nAm I doing something fundamentally wrong here?"

In [ ]:
first_five['text'][4]

"\n\n  This is the latest from UPI \n\n     Foreign Ministry spokesman Ferhat Ataman told journalists Turkey was\n     closing its air space to all flights to and from Armenia and would\n     prevent humanitarian aid from reaching the republic overland across\n     Turkish territory.\n\n  \n   Historically even the most uncivilized of peoples have exhibited \n   signs of compassion by allowing humanitarian aid to reach civilian\n   populations. Even the Nazis did this much.\n\n   It seems as though from now on Turkey will publicly pronounce \n   themselves 'hypocrites' should they choose to continue their\n   condemnation of the Serbians.\n\n\n\n--"

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (19).ipynb
    
</div>

In [ ]:
document_topics = lda.transform(X)

for i in range(5):
    print("-----")
    dominant_topic_index = np.argmax(document_topics[i])
    print(f"Document {i+1} Topic Assignment: Topic {dominant_topic_index}")
    print()
    print(df["text"].iloc[i])
    

-----
Document 1 Topic Assignment: Topic 6

You know, I was reading 18 U.S.C. 922 and something just did not make 
sence and I was wondering if someone could help me out.

Say U.S.C. 922 :

(1) Except as provided in paragraph (2), it shall be unlawful for
any person to transfer or possess a machinegun.

 Well I got to looking in my law dictionary and I found that a "person" 
might also be an artificial entity that is created by government 
and has no rights under the federal constitution. So, what I 
don't understand is how a statute like 922 can be enforced on 
an individual. So someone tell me how my government can tell
me what I can or cannot possess. Just passing a law 
does not make it LAW. Everyone knows that laws are constitional
until it goes to court. So, has it ever gone to court, not
just your run of the mill "Ok I had it I am guilty, put me in jail"

Has anyone ever claimed that they had a right to possess and was told
by the Supreme Court that they didn't have that right?


Generally, our documents seem to be under the correct topics. Document 3's assignment to Topic 0 is a bit strange, since Topic 0 seems to revolve around religion. However, Document 3's content is a bit more abstract and does not seem to contain much talk of graphics anyways, so this is understandable. The other 4 documents seem to be under appropriate topics.   

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (15).ipynb
    
</div>

In [ ]:
print(document_topics[0], document_topics[1], document_topics[2], 
      document_topics[3], document_topics[4])

[0.91792255 0.0048727  0.07251588 0.00468886] [0.00368813 0.98887899 0.00376892 0.00366397] [0.00502992 0.00530702 0.83905815 0.15060491] [0.00928782 0.9717597  0.00996268 0.0089898 ] [0.98160245 0.00609432 0.00623438 0.00606885]


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (23).ipynb
    
</div>

In [ ]:
document_topics = lda_model.transform(X)

for doc_idx, topic_dist in enumerate(document_topics[:5]):
    print(f"Document {doc_idx} Topic Distribution: {topic_dist}")

Document 0 Topic Distribution: [0.0049736  0.00508882 0.98494165 0.00499593]
Document 1 Topic Distribution: [0.20989649 0.24565277 0.00362345 0.54082728]
Document 2 Topic Distribution: [0.00514927 0.66194901 0.0720806  0.26082112]
Document 3 Topic Distribution: [0.01164391 0.01233411 0.09765885 0.87836313]
Document 4 Topic Distribution: [0.00804408 0.00814161 0.97574605 0.00806826]


Document 0: This document relates to the second topic (politics) very heavily

Document 1: Most likely related to topic 3 the most (Computing and Technology), however, it has a relatively weak probability for this categorization

Document 2: This document relates to the first topic (philosophy and religion). It's probability is relatively weak compared to other documents but still higher than document 1

Document 3: Related to topic 3 (computing and technology) very heavily

Document 4: This document relates to the second topic (politics) very heavily

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (40).ipynb
    
</div>

citation: code from/reference to cpsc 330 lecture 17

In [ ]:
print(document_topics[0:5])
print(df["text_pp"][0:5])

[[0.00466079 0.79523864 0.00471635 0.19538421]
 [0.46949436 0.29534213 0.0037875  0.231376  ]
 [0.00479165 0.00483106 0.17724669 0.81313059]
 [0.00893626 0.35154545 0.29518598 0.34433231]
 [0.00589146 0.1444478  0.00593581 0.84372494]]
0    know read u.s.c. 922 sence wonder help u.s.c. ...
1    bad question ref list algorithm think bit hard...
2    actuallay hand support idea have newsgroup asp...
3    follow problem bug appreciate help create wind...
4    late upi foreign ministry spokesman ferhat ata...
Name: text_pp, dtype: object


The document topic assignments vary for the topic groups. Seems like our fourth topic is more ambiguous. 

<div class="alert alert-warning">

hw7_sol (31).ipynb
    
</div>

In [ ]:
document_topics = lda_model.transform(X)

for i in range(5):
    print("Document " + str(i + 1))
    print(document_topics[i])

Document 1
[0.00181855 0.00181864 0.00181854 0.70069978 0.00181836 0.28475231
 0.00181836 0.00181852 0.00181824 0.0018187 ]
Document 2
[0.00140885 0.00140867 0.00140875 0.00140886 0.46708967 0.00140857
 0.00140868 0.40253992 0.00140881 0.12050923]
Document 3
[0.27706948 0.00178616 0.00178608 0.00178622 0.00178587 0.0017858
 0.00178593 0.001786   0.00178602 0.70864244]
Document 4
[0.00333409 0.00333394 0.62183542 0.00333422 0.0033337  0.00333347
 0.0033334  0.00333356 0.00333357 0.35149464]
Document 5
[0.03495029 0.0022228  0.00222252 0.66844391 0.00222239 0.10681574
 0.0022224  0.00222284 0.17645441 0.00222269]


It seems that document 1 is most likely to belong in topic 4, document 2 in either topic 5 or 8, document 3 in topic 10, document 4 in topic 3, and document 5 in topic 4

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (1).ipynb
    
</div>

In [ ]:
topic_assignment = [topics[key] for key in dominant_topic]

final_df = pd.concat([df, pd.DataFrame(topic_assignment, columns=["topic_assignment"])], axis=1)

final_df.head()

text  target  \
0  You know, I was reading 18 U.S.C. 922 and some...       6   
1  \n\n\nIt's not a bad question: I don't have an...       1   
2  \nActuallay I don't, but on the other hand I d...       1   
3  The following problem is really bugging me,\na...       2   
4  \n\n  This is the latest from UPI \n\n     For...       7   

             target_name                                            text_pp  \
0     talk.politics.guns  know read u.s.c. sence wonder help u.s.c. prov...   
1          comp.graphics  bad question ref list algorithm think bit hard...   
2          comp.graphics  actuallay hand support idea have newsgroup asp...   
3         comp.windows.x  follow problem bug appreciate help create wind...   
4  talk.politics.mideast  late upi foreign ministry spokesman ferhat ata...   

        topic_assignment  
0  Religion and Politics  
1              Computers  
2              Computers  
3              Computers  
4             World News

2. The document assignment has seemed to work okay. It has grouped the comp.graphics and comp.windows.x targets into the computers topic. The relgion and politics topic as well as the world news topics also seem appropriate for df[0] and df[4]. This is quite similar to what I have predicted and the topic assignments seem to pair quite well with the original and preprocessed texts.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (35).ipynb
    
</div>

In [ ]:
for i in range(0, 5):
    print(f"the topic association for {df['text_pp'][i]} is {topics[i]}")

the topic association for know read 18 U.S.C. 922 
 sence wonder help 

 U.S.C. 922 

 1 provide paragraph 2 shall unlawful 
 person transfer possess machinegun 

  get look law dictionary find person 
 artificial entity create government 
 right federal constitution 
 understand statute like 922 enforce 
 individual tell government tell 
 possess pass law 
 law know law constitional 
 go court go court 
 run mill ok guilty jail 

 claim right possess tell 
 Supreme Court right 


 is [0.26226783 0.61223871 0.11452105 0.00219616 0.00219299 0.00219299
 0.00219686 0.00219343]
the topic association for 


 bad question ref list algorithm 
 think bit hard 

 1 point find plane define 
 circle lie algorithm 

 2 find center   circle line pass center 
 perpendicular plane point pass center 
 sphere 

 3 repeat unused point original point 
 give different line pass sphere 
 origin interection center sphere 

 4 radius easy compute distance center 
 original point 

 leave math workable algori

The topic assignment is not perfect as there is some overlap between some topics and there are some words that could be considered as two different topics. The vaguer categories have random words that could be included in any other category, which suggests that the topic association isn't great. 

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (48).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f"DOCUMENT {i}: ")
    print(df.iloc[i]["text"])
    print(f"It has topics {document_topics[i]}\n")

DOCUMENT 0: 
You know, I was reading 18 U.S.C. 922 and something just did not make 
sence and I was wondering if someone could help me out.

Say U.S.C. 922 :

(1) Except as provided in paragraph (2), it shall be unlawful for
any person to transfer or possess a machinegun.

 Well I got to looking in my law dictionary and I found that a "person" 
might also be an artificial entity that is created by government 
and has no rights under the federal constitution. So, what I 
don't understand is how a statute like 922 can be enforced on 
an individual. So someone tell me how my government can tell
me what I can or cannot possess. Just passing a law 
does not make it LAW. Everyone knows that laws are constitional
until it goes to court. So, has it ever gone to court, not
just your run of the mill "Ok I had it I am guilty, put me in jail"

Has anyone ever claimed that they had a right to possess and was told
by the Supreme Court that they didn't have that right?



It has topics [0.77106703 0.

2. 

Document 0 is said to be most related to topic 0, which we said is war. it lines up

Document 1 is said to be most related to topic 2, which we said is computer basics. it lines up

Document 2 is said to be related to topic 1 and topic 2, which we said is Faith and computer basics. it lines up more with computer basics

Document 3 is said to be most related to topic 3, which we said is Software. it lines up

Document 4 is said to be most related to topic 0, which we said is war. it lines up



<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (27).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f"Text {i + 1}: {topics[i]}")

print(df["text"].iloc[:5])

print(topic_names)


Text 1: [0.0060457  0.83948688 0.15446742]
Text 2: [0.00526126 0.00538506 0.98935368]
Text 3: [0.00609212 0.00615734 0.98775054]
Text 4: [0.97698119 0.01229339 0.01072542]
Text 5: [0.00822215 0.98331115 0.0084667 ]
0    You know I was reading 18 U.S.C. 922 and somet...
1    It's not a bad question I don't have any refs ...
2    Actuallay I don't but on the other hand I don'...
3    The following problem is really bugging meand ...
4      This is the latest from UPI      Foreign Min...
Name: text, dtype: object
computers, sports, religion


Text 1 was assigned to religion. 
Text 2 was assigned to computers. 
Text 3 was assigned to computers. 
Text 4 was assigned to computers. 
Text 5 was assigned to religion. 


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (39).ipynb
    
</div>

In [ ]:
doc_vect = vect.transform(df['text_pp'])
topic_dist = lda.transform(doc_vect)

In [ ]:
for i in range(5):
    text = df.iloc[i]['text']
    i=i+1
    print("Document", i)
    print(text)

    print("Topic Distribution")
    for j, prob in enumerate(topic_dist[i]):
        print(f"~{prob * 100:.2f}% topic {j}")
    print()

_Document 1 is most associated with topics 5 and 8 with less contributions of topic 4. Document 2 is most associated with topic 1 and some contributions of topic 8. Document 3 is most associated with topics 4 and 8 and some contributions of topic 7. Document 4 is only mainly associated with topic 4. Document 5 is most associated with topic 8 and some contributions of topics 1 and 7.  

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (44).ipynb
    
</div>

In [ ]:
doc_topic_distributions = lda.transform(X)

first_five_docs = doc_topic_distributions[:5]

# chatgpt 4.0 for for loop
for i in range(5):
    distribution = first_five_docs[i]
    print(f"Document {i} Topic distribution: {distribution}")
    print(f"{df.loc[i, 'text_pp']}\n")

Document 0 Topic distribution: [0.13646702 0.46711131 0.00265059 0.38581842 0.0026508  0.00265383
 0.00264804]
know read u.s.c. 922 sence wonder help u.s.c. 922 provide paragraph shall unlawful person transfer possess machinegun get look law dictionary find person artificial entity create government right federal constitution understand statute 922 enforce individual tell government tell possess pass law law know law constitional go court go court run mill guilty jail claim right possess tell supreme court right

Document 1 Topic distribution: [0.00201471 0.00201766 0.00201511 0.98790013 0.00201929 0.00201613
 0.00201697]
bad question ref list algorithm think bit hard point find plane define circle lie algorithm find center circle line pass center perpendicular plane point pass center sphere repeat unused point original point give different line pass sphere origin interection center sphere radius easy compute distance center original point leave math workable algorithm alternate method

Given the topics:
- Topic 0: Firearms
- Topic 1: General Discussion
- Topic 2: Religion
- Topic 3: Atheism/Questioning
- Topic 4: Computer Science
- Topic 5: Middle Eastern Discussion
- Topic 6: Sports


**Document 0:**
Topic distribution: [0.13646702 0.46711131 0.00265059 0.38581842 0.0026508  0.00265383  0.00264804] 

The topic distribution mainly concerns 14% Firearms, 47% General Discussion and 38% Questioning words. This seems pretty representative under a BoW interpretation as the word distribution seems to follow a similar magnitude.


**Document 1:**
Topic distribution: [0.00201471 0.00201766 0.00201511 0.98790013 0.00201929 0.00201613 0.00201697]

The topic distribution mainly concerns 98% Computer Science. This seems pretty representative given its content.


**Document 2:**
Topic distribution: [0.0027493  0.30803689 0.00275238 0.53763222 0.14332366 0.00275185 0.00275371]

The topic distribution mainly concerns 30% General Discussion, 53% Questioning Language, and 14% Computer Science. This seems pretty representative given its content as it seems like an email discussion.


**Document 3:**
Topic distribution: [0.15149818 0.00494263 0.0049366  0.00495603 0.47107632 0.35765998 0.00493025]

The topic distribution mainly concerns 15% General Discussion, 47% Computer Science, and 35% Middle Eastern Discussion. I don’t see where the middle eastern comes in as this seems to be almost purely computer science. 


**Document 4:**
Topic distribution:[0.00332487 0.00333917 0.00335736 0.00335298 0.1738728 0.80942674 0.00332609]

The topic distribution mainly concerns 17% Computer Science and 80% Middle Eastern Discussion. I do not see where Computer Science comes in however it correctly identified the Middle Eastern discussion.

Overall the model did a fairly good job identifying categories, however it seemed to mix middle eastern with computer science towards the end. 


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (11).ipynb
    
</div>

In [ ]:
topic_assignments = [np.argmax(topic) for topic in document_topics[:5]]
topic_assignments

[2, 1, 2, 1, 0]

In [ ]:
df[["target_name"]].head(5)

target_name
0     talk.politics.guns
1          comp.graphics
2          comp.graphics
3         comp.windows.x
4  talk.politics.mideast

The model incorrectly assigns topic `religion` to document 0.\
The model correctly assigns topic `computers` to document 1.\
The model incorrectly assigns topic `religion` to document 2.\
The model correctly assigns topic `computers` to document 3.\
The model correctly assigns topic `politics` to document 4.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (46).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(document_topics[i])

[0.00501964 0.00498656 0.98492757 0.00506624]
[0.00366609 0.00369407 0.1559993  0.83664054]
[0.00487997 0.00497126 0.69445644 0.29569233]
[0.00897028 0.00876083 0.05805558 0.92421331]
[0.7663498  0.00596384 0.22153447 0.00615189]


The array shows how likely a document is to be about a topic. Document 1 is most likely about topic 2 (98%, religion), doc 2 is likely about topic 3 (83%, computer), doc 3 about topic 2 (69%, religion), doc 4 about topic 3 (92%, computer), and doc 5 about topic 0 (76%, armanian genocide/war).

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (25).ipynb
    
</div>

In [ ]:
topics = [
    'alt.atheism',
    'comp.graphics',
    'comp.windows.x',
    'rec.sport.baseball',
    'rec.sport.hockey',
    'soc.religion.christian',
    'talk.politics.guns',
    'talk.politics.mideast'
]
# Create a dictionary mapping each index to the corresponding topic
topic_dict = {i: topic for i, topic in enumerate(topics)}
query_n = 5

for n in range(query_n):
    topic = np.argmax(document_topics[n])
    print("The topic association for \"document %d\" is: topic %d - %s" % (n, topic, topic_dict[topic]))


The topic association for "document 0" is: topic 2 - comp.windows.x
The topic association for "document 1" is: topic 0 - alt.atheism
The topic association for "document 2" is: topic 2 - comp.windows.x
The topic association for "document 3" is: topic 2 - comp.windows.x
The topic association for "document 4" is: topic 7 - talk.politics.mideast


2. The model accurately identified the topics for documents 2, 3, and 4 but made errors with documents 0 and 1. Document 0 was incorrectly classified as 'comp.windows.x' instead of the correct 'talk.politics.guns', and document 1 was misclassified as 'alt.atheism' instead of 'comp.windows.x'. This indicates the model's strong performance in identifying computer-related topics but shows a need for improvement in accurately distinguishing between other non-technical topics.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (29).ipynb
    
</div>

In [ ]:
topic_distributions = lda_model.transform(X_vectorized)

most_likely_topics = topic_distributions.argmax(axis=1)[:5]

df_first_five = df.head().copy()
df_first_five['Most Likely Topic'] = most_likely_topics + 1  

pd.set_option('display.max_colwidth', None)

print(df_first_five[['text', 'Most Likely Topic']])

### Analysis of Snippets to Topics

1. **Snippet 1 (Topic 3: Religion and Society)**
   - The discussion about laws and rights might seem better suited to a legal or political category, but it's placed here possibly due to the absence of a more fitting topic. The topic's focus on societal structures could loosely encompass legal debates.

2. **Snippet 2 (Topic 6: Software Development)**
   - Describes a technical procedure for finding a sphere's center, fitting well with topics related to programming or mathematical computations, aligning with "Software Development".

3. **Snippet 3 (Topic 8: Philosophy and Belief Systems)**
   - Discusses preferences for structuring discussion groups, which is somewhat related to personal beliefs and systems of organizing information, hence the philosophical aspect.

4. **Snippet 4 (Topic 6: Software Development)**
   - Talks about specific software behavior and event handling, clearly aligning with "Software Development".

5. **Snippet 5 (Topic 10: Historical and Ethnic Conflicts)**
   - Mentions Turkey's actions towards Armenia, touching on historical and geopolitical issues, making "Historical and Ethnic Conflicts" a relevant topic.


<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (37).ipynb
    
</div>

In [ ]:

topic_labels = ["Topic 0", "Topic 1", "Topic 2", "Topic 3", "Topic 4"]
topic_assignments = lda.transform(X[:5])
for i, topic_probs in enumerate(topic_assignments):
    print(f"Document {i + 1} Topic Distribution:", topic_probs)
    dominant_topic = topic_probs.argmax()
    if dominant_topic < len(topic_labels):  # Ensure dominant_topic is within range
        print(f"Document {i + 1} Dominant Topic:", dominant_topic, "-", topic_labels[dominant_topic])
    else:
        print(f"Document {i + 1} Dominant Topic:", dominant_topic, "- No label available")
    print()




Document 1 Topic Distribution: [0.00350942 0.00350021 0.00350901 0.00351063 0.98597073]
Document 1 Dominant Topic: 4 - Topic 4

Document 2 Topic Distribution: [0.00287662 0.59370846 0.39766549 0.00284505 0.00290438]
Document 2 Dominant Topic: 1 - Topic 1

Document 3 Topic Distribution: [0.00392924 0.00391409 0.20054086 0.00386212 0.78775369]
Document 3 Dominant Topic: 4 - Topic 4

Document 4 Topic Distribution: [0.00749301 0.00744341 0.07970162 0.00742656 0.89793539]
Document 4 Dominant Topic: 4 - Topic 4

Document 5 Topic Distribution: [0.00474522 0.00470917 0.00471451 0.26615698 0.71967412]
Document 5 Dominant Topic: 4 - Topic 4



<!-- END QUESTION -->
we'll analyze the document topic assignments and assess whether they align with our expectations or provide meaningful insights into the content of the documents. If the topic assignments appear coherent and relevant to the content of the documents, it indicates that the LDA model has successfully captured the underlying topics in the corpus. However, if the assignments seem random or unrelated to the document content, it suggests that the model may not be effectively capturing the latent topics
<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (13).ipynb
    
</div>

In [ ]:
for i in range(5):
    print(f"Document {i}: {df['text_pp'][i]}")
    print(f"Document {i} assignment: {document_topics[i].argmax()}")

Document 0: know read sence wonder help provide paragraph shall unlawful person transfer possess machinegun get look law dictionary find person artificial entity create government right federal constitution understand statute like enforce individual tell government tell possess pass law law know law constitional go court go court run mill ok guilty jail claim right possess tell Supreme Court right
Document 0 assignment: 2
Document 1: bad question ref list algorithm think bit hard point find plane define circle lie algorithm find center circle line pass center perpendicular plane point pass center sphere repeat unused point original point give different line pass sphere origin interection center sphere radius easy compute distance center original point leave math workable algorithm alternate method pair point plane form perpendicular bisector line segment pair contain center sphere pair form plane intersect point easy implement
Document 1 assignment: 0
Document 2: Actuallay hand support

The document topic assignment of the model seems to be quite accurate. The topics assigned to the documents are relevant to the content of the documents.

<!-- END QUESTION -->

<br><br><br><br>

<div class="alert alert-warning">

hw7_sol (3).ipynb
    
</div>

In [ ]:
# sorting = np.argsort(lda.components_, axis=1)[:, ::-1]
feature_names = newsgroups_train.target_names

In [ ]:
import mglearn

mglearn.tools.print_topics(
    topics=range(3),
    feature_names=feature_names,
    sorting=sorting,
    topics_per_chunk=5,
    n_words=10,
)

<!-- END QUESTION -->

<br><br><br><br>